Multihomed support for client socket in snet

juagargi opened the pull request:
This PR adds a basic multihomed support for sockets created using `snet`.
It allows to create SCION `snet` sockets without binding them to a local address.
The changes should not affect any previous code that created sockets binding them to a local address.
Two use cases are relevant, and used as examples:
- A client opens a socket to send a message to a given remote, not needing to bind the socket to any interface/address in particular.
- A server opens a socket to listen to any connection, not bound to any interface/address in particular.

In the first case, the local address is found via (a syscall to the kernel) creating a regular UDP socket with the specified destination, letting the kernel check the routing table and deciding what local IP address the socket will have.

In the second case, the packet received contains the last hop, which is used to create a regular UDP socket, similar to the case above. This time, the socket is not bound to the local address, and the local address is used only to allow typical constructs in the server such as
```go
// Read data from the client.
n, remoteAddr, err := conn.ReadFrom(buff)
...
// Reply to the client.
n, err = conn.WriteTo([]byte(buff), remoteAddr)
```

In both cases no network traffic is created, and only the syscalls are needed to communicate with the kernel.

Additionally, there is a mechanism that reduces the amount of syscalls needed to open an unbound `snet` socket, and it is based on caching previous syscalls to the kernel.
If we denote `outbound(remote)` the function that returns the local IP address for a given remote address (the next hop of a SCION path), we keep a map of these results. If the cached map doesn't contain an entry for a given `remote`, the regular `outbound(remote)` is executed, thus issuing syscalls.

The cache is valid only for as long as the local addresses do not change, or there is a change in the routing table. We assume that a change of the routing table will be always accompanied by a change on the local interfaces' configuration, be it in the form of their local addresses, or creation/removal of some of these interfaces. If such an event is detected, the whole cache is cleared.

The detection of the changes should be done via e.g. `netlink`, but this PR is not using it, instead relying on the much simpler approach of querying the local interface state every T seconds, where T is defined here as 1 second.
This implies that in the case of a change of interface local address, or creation/deletion of an interface, there could be traffic being sent using a wrong local address (for at most T seconds).

This PR enables binding of SCION sockets to unspecified IP, with the following uncovered cases (i.e., clients/servers that rely on this feature may lose connectivity, drop packets):
- Routing changes on the endhost that do not imply interface changes.
- In case of a change of interface local address, or creation/deletion of an interface, there could be traffic being sent using a wrong local address (for at most T seconds).

In order to tackle this shortcoming one should implement a different solution using e.g. netlink.

In the case of using a dispatcher to obtain or send the packets, although the `snet` socket would NOT require to be unbound to any local address, this PR should not introduce any _visible_ effects.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/juagargi/scion/53)
<!-- Reviewable:end -->

juagargi committed:
New test using tiny topo.

juagargi committed:
Refactor test.

juagargi committed:
First phase of multihoming test.

juagargi committed:
Temporary changes.

juagargi committed:
Multihomed behavior through UDP socket creation.

Via the creation of a UDP socket with destination the next hop,
those snet sockets that were created without binding them to a local
address, can now be used regularly.
There is a additional cache that allows skipping the UDP socket creation
if the address to dial to has been previously seen.
There is a recurrent refresh function running in the background that
checks that the host interfaces' local addresses didn't change, and if
they did, clears the cache.

juagargi committed:
Limit cache size.

juagargi committed:
Make multihomed tests more robust to concurrent running.

juagargi committed:
Fix gateway/dataplane TestNoLeak by allowing to stop the continous state getter.

JordiSubira reviewed (commented):
Currently this change tackles routing changes that imply a change in the interface/address list, i.e., the 2 cases you describe at `interfaces.go`. You should **explicitly mention** that this change does not cover routing changes on the endhost that do not imply interface changes. In other words, the source address on the SCION packet may have a wrong value if this ever happens. 

Just to clarify the scenario I am think of:
-   **Interface 0** with IP address `A.A.A.A`
-   **Interface 1** with IP address `B.B.B.B`
-   Destination `X.X.X.X`
Initially our routing table uses interface A.A.A.A:
```
default via A.A.A.A dev eth0 proto dhcp metric 100 
default via B.B.B.B dev eth1 proto dhcp metric 200
```
 Due to some network change due to a dynamic protocol or manual configuration the endhost routing table gets updated to
```
default via A.A.A.A dev eth0 proto dhcp metric 100 
default via B.B.B.B dev eth1 proto dhcp metric 200
X.X.X.X/16 via B.B.B.B dev eth1
```
However, the internal cache doesn't get updated based on the description. Similarly, I haven't checked whether the code handles the fact that the interface is down although still appears on the list with the same assigned IP address.

Perhaps this is not relevant for the use case this PR is trying to initially support, but it is important to mention in case this PR is ever used as referenced to merge in scionproto.

> @JordiSubira reviewed 8 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/juagargi/scion/53#-:-OZd7hMdA48UUSH5lKGE:bai1r6)* status: 8 of 14 files reviewed, 6 unresolved discussions

___
*[`pkg/snet/multihomed/outbound_addr.go` line 26 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcXaUIA10fEKxT_Mx4:-OZcXaUIA10fEKxT_Mx5:b-cqotjc) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/multihomed/outbound_addr.go#L26)):*
> ```Go
> // It relies on a previously populated table that maps remote addresses to egress addresses.
> // If the remote is not present, it is added.
> func OutboundIP(raddr *net.UDPAddr) (net.IP, error) {
> ```

Nit: change parameter name to `nextHop` normally we use raddr as endhost remote address.
___
*[`pkg/snet/multihomed/outbound_addr.go` line 31 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcH0ez6IrUfa5-u2uK:-OZcH0f-97K2HRLUAwfw:b-erytyc) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/multihomed/outbound_addr.go#L31)):*
> ```Go
> 	defer muRemoteToEgress.RLocker().Unlock()
> 
> 	remote, _ := netip.AddrFromSlice(raddr.IP)
> ```

It will panic if raddr is nil. I guess most on the inputs are constructed properly by previous library calls, but I am normally in favor of handling this returning an error than just panicking...
___
*[`pkg/snet/multihomed/outbound_addr.go` line 57 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZd3vSs18mFiBvifKhs:-OZd3vSt5Cnr4z5TY5kx:b-mbz5ky) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/multihomed/outbound_addr.go#L57)):*
> ```Go
> 	}
> 	return eg, nil
> }
> ```

The lock logic here seems a bit contrived, could we simplify for sth like this:
```
func OutboundIP(raddr *net.UDPAddr) (net.IP, error) {
    remote, _ := netip.AddrFromSlice(raddr.IP)
    
    muRemoteToEgress.RLock()
    egress, ok := remoteToEgress[remote]
    muRemoteToEgress.RUnlock()
    
    if ok {
        return net.IP(egress.AsSlice()), nil
    }

    eg, err := dialRemote(raddr)
    if err != nil {
        return nil, err
    }
    
    egress, _ = netip.AddrFromSlice(eg)
    muRemoteToEgress.Lock()
    if len(remoteToEgress) < MaxAllowedCacheSize {
        remoteToEgress[remote] = egress
    }
    muRemoteToEgress.Unlock()
    
    return eg, nil
}
``
```




_Code quote:_
```Go
	// Check if the table contains an entry.
	muRemoteToEgress.RLock()
	defer muRemoteToEgress.RLocker().Unlock()

	remote, _ := netip.AddrFromSlice(raddr.IP)
	egress, ok := remoteToEgress[remote]
	if ok {
		return net.IP(egress.AsSlice()), nil
	}

	// Not found, find it and add it. The dialing involves a syscall, but no network traffic.
	eg, err := dialRemote(raddr)
	if err != nil {
		return nil, err
	}
	egress, _ = netip.AddrFromSlice(eg)

	// Check if our cache is not too big already.
	if len(remoteToEgress) < MaxAllowedCacheSize {
		// Beware of the RWMutex, it's read-locked already.
		muRemoteToEgress.RUnlock()
		muRemoteToEgress.Lock()
		// Check again to avoid race conditions. Could skipped it if exact size is not important.
		if len(remoteToEgress) < MaxAllowedCacheSize {
			remoteToEgress[remote] = egress
		}
		muRemoteToEgress.Unlock()
		muRemoteToEgress.RLock() // because we always read-unlock in this function.
	}
	return eg, nil
}
```
___
*[`pkg/snet/conn.go` line 79 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcFNJU8Siv4wy_fKE9:-OZcFNJU8Siv4wy_fKEA:b-xubke5) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/conn.go#L79)):*
> ```Go
> 	// If acting as a client, find the local address of the interface used to reach
> 	// the NextHop toward the remote.
> 	if local.Host.IP.IsUnspecified() && o.remote != nil {
> ```

Keep the non-nil check on `local.host` as a safeguard to panic. 
___
*[`pkg/snet/conn.go` line 87 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcTEfS1UTmzH4AqJiv:-OZcTEfS1UTmzH4AqJiw:b-jd9k9l) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/conn.go#L87)):*
> ```Go
> 				"remote", o.remote)
> 		}
> 		local.Host.IP = localIP
> ```

Perhaps I am not understanding this completely, but it seems that you fix the local address upon creation for clients. What happens for long-lived connections in which the nextHop address changes. One can use snet.WriteTo(msg, remote) por a remote that has changed the path, for instance what pan.Conn does. In this case, the local.Host.IP for the client would need to change, wouldn't it?
___
*[`pkg/snet/reader.go` line 96 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcJqjE2_808LP1zdxJ:-OZcJqjE2_808LP1zdxK:b2nt4r2) ([raw file](https://github.com/juagargi/scion/blob/150bf5830b5efc85e77d81545aff2d58f76566e4/pkg/snet/reader.go#L96)):*
> ```Go
> 	// we would like to replace this logic (e.g., using IP_PKTINFO, with its caveats).
> 	pktAddrPort := netip.AddrPortFrom(pkt.Destination.Host.IP(), udp.DstPort)
> 	if c.local.IA != pkt.Destination.IA ||
> ```

I am a bit reluctant on just removing, this check. The reason is that we decouple the underlay from the SCION layer completely. We should clearly document this for applications so that library users understand the implications. More concretely, that SCION_SRC_IP/PORT in the packet doesn't correspond to the underlay for which the packet was received.


<!-- Sent from Reviewable.io -->

juagargi committed:
Do not obtain local address at Conn creation, do it at WriteTo.

juagargi committed:
More efficient & clear mutex locking in OutboundIP.

juagargi reviewed (commented):
Correct, I thought it was clear in the PR description, but I found no documentation on the code itself.
Now I have added some sentences in the `NewCookedConn` documentation clarifying that the returned `Conn` reacts to routing changes only if they come with a change in the local network interface set.

> *[Reviewable](https://reviewable.io/reviews/juagargi/scion/53#-:-OZn-eQm-Obt53_Fr5Mi:beqn00s)* status: 6 of 14 files reviewed, 5 unresolved discussions (waiting on @JordiSubira)

___
*[`pkg/snet/conn.go` line 79 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcFNJU8Siv4wy_fKE9:-OZixjW31OL3sqeLfBwQ:b-qgsfjv) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/conn.go#L79)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

Keep the non-nil check on `local.host` as a safeguard to panic.
</blockquote></details>

We have a type assertion right above setting local.Host to a `*net.UDPAddr`, AFAIK it will already panic or set `local.Host` to a non nil value, is that correct? If so, do you want me to check `pconn.LocalAddr` before is type asserted?
___
*[`pkg/snet/conn.go` line 87 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcTEfS1UTmzH4AqJiv:-OZiyWh_C92lzHPm__OK:b-12snkj) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/conn.go#L87)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

Perhaps I am not understanding this completely, but it seems that you fix the local address upon creation for clients. What happens for long-lived connections in which the nextHop address changes. One can use snet.WriteTo(msg, remote) por a remote that has changed the path, for instance what pan.Conn does. In this case, the local.Host.IP for the client would need to change, wouldn't it?
</blockquote></details>

What you say is correct, I think this code may not be very robust: when the connection is created via `NewCookedConn` the socket gets the local IP to that that reaches the next hop at this moment, but after some time, the next hop maybe different. In this case, the local IP maybe different too.
Since `PacketConn` offers `WriteTo` with the address of the next hop specified as a parameter, the construction of the socket in `NewCookedConn` should not attempt to get the local IP address at that moment in time, but always leave it to the `WriteTo` logic, where the next hop address is known.
___
*[`pkg/snet/reader.go` line 96 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcJqjE2_808LP1zdxJ:-OZn2EaR9CczLO1LRNpr:b-k3zl3c) ([raw file](https://github.com/juagargi/scion/blob/150bf5830b5efc85e77d81545aff2d58f76566e4/pkg/snet/reader.go#L96)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

I am a bit reluctant on just removing, this check. The reason is that we decouple the underlay from the SCION layer completely. We should clearly document this for applications so that library users understand the implications. More concretely, that SCION_SRC_IP/PORT in the packet doesn't correspond to the underlay for which the packet was received.
</blockquote></details>

Understood. My rationale for removing it is to decouple routing/packet delivery from the snet library if possible, i.e. "if the packet arrived here, there must have been a reason for it to arrive". Please let me know if you think this is not appropriate at this stage of the scionproto development and on the other hand if this seems okay, where I should write some clarification on how UDP packets are used as underlay for SCION packets.
___
*[`pkg/snet/multihomed/outbound_addr.go` line 31 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcH0ez6IrUfa5-u2uK:-OZj0uIi5HL80xX99Aw6:b-896fix) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/multihomed/outbound_addr.go#L31)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

It will panic if raddr is nil. I guess most on the inputs are constructed properly by previous library calls, but I am normally in favor of handling this returning an error than just panicking...
</blockquote></details>

Done.
___
*[`pkg/snet/multihomed/outbound_addr.go` line 57 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZd3vSs18mFiBvifKhs:-OZn42bY59Zx900800S6:b6ku427) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/multihomed/outbound_addr.go#L57)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

The lock logic here seems a bit contrived, could we simplify for sth like this:
```
func OutboundIP(raddr *net.UDPAddr) (net.IP, error) {
    remote, _ := netip.AddrFromSlice(raddr.IP)
    
    muRemoteToEgress.RLock()
    egress, ok := remoteToEgress[remote]
    muRemoteToEgress.RUnlock()
    
    if ok {
        return net.IP(egress.AsSlice()), nil
    }

    eg, err := dialRemote(raddr)
    if err != nil {
        return nil, err
    }
    
    egress, _ = netip.AddrFromSlice(eg)
    muRemoteToEgress.Lock()
    if len(remoteToEgress) < MaxAllowedCacheSize {
        remoteToEgress[remote] = egress
    }
    muRemoteToEgress.Unlock()
    
    return eg, nil
}
``
```

</blockquote></details>

Correct, your version is not only more elegant, is also locking less amount of time. I'm using it.


<!-- Sent from Reviewable.io -->

JordiSubira reviewed (commented):
To me it was not clear on the description. Aside from the in-line comment to the Conn constructor, can you add in the PR description:
> This PR enables binding of SCION sockets to unspecified IP, with the following **uncovered cases** (i.e., clients/servers that rely on this feature may lose connectivity, drop packets):
> - Routing changes on the endhost that do not imply interface changes.
> - In case of a change of interface local address, or creation/deletion of an interface, there could be traffic being sent using a wrong local address (for at most T seconds).

>In order to tackle this shortcoming one should implement a different solution using e.g. `netlink`.

> *[Reviewable](https://reviewable.io/reviews/juagargi/scion/53#-:-OZnSgcn3VjAV3m9cszW:b-k21dl9)* status: 6 of 14 files reviewed, 3 unresolved discussions (waiting on @juagargi)

___
*[`pkg/snet/conn.go` line 79 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcFNJU8Siv4wy_fKE9:-OZnCrPv6DtX5P-4Q5lq:b6n9kcu) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/conn.go#L79)):*
<details><summary><i>Previously, juagargi (Juan A. Garcia Pardo) wrote…</i></summary><blockquote>

We have a type assertion right above setting local.Host to a `*net.UDPAddr`, AFAIK it will already panic or set `local.Host` to a non nil value, is that correct? If so, do you want me to check `pconn.LocalAddr` before is type asserted?
</blockquote></details>

Yes, you are right I missed that one, don't think it needs to be checked then.
___
*[`pkg/snet/reader.go` line 96 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcJqjE2_808LP1zdxJ:-OZnDL7KD5ECllI5AG1N:b-vsy7rp) ([raw file](https://github.com/juagargi/scion/blob/150bf5830b5efc85e77d81545aff2d58f76566e4/pkg/snet/reader.go#L96)):*
<details><summary><i>Previously, juagargi (Juan A. Garcia Pardo) wrote…</i></summary><blockquote>

Understood. My rationale for removing it is to decouple routing/packet delivery from the snet library if possible, i.e. "if the packet arrived here, there must have been a reason for it to arrive". Please let me know if you think this is not appropriate at this stage of the scionproto development and on the other hand if this seems okay, where I should write some clarification on how UDP packets are used as underlay for SCION packets.
</blockquote></details>

If we are going to allow this, the check will even be better at `snet.OpenRaw()`.

This is not about the packet arriving, is about the information on the packet being meaningful, i.e., the src address on the SCION layer being decoupled from the underlay. So far, this wasn't the case for snet (we didn't allow for Unspecified address, not even before the dispatcher-less change). 

This may have implications, not all of them completely obvious perhaps at the moment, but I can try to come up with an "artifact" example: A SCION application that listens on 0.0.0.0 wants to display how many packets is receiving per interface. One could think that I can use pkt.Destination.Host (populated at snet.PacketConn.ReadFrom(pkt)) and infer from that address the interface for which I am receiving packets. This is not correct, because one could send a SCION packet to the underlay X.X.X.X with SCION destination address Y.Y.Y.Y. Of course in this model, "the attacker" is local (meaning within the AS). 

I am not saying we have to care about this "misuse case" which as I stated is a bit of an artifact, what I am saying is that we ever have this change on production, this has to be documented. My suggestion is even having it documented at snet.OpenRaw() because the example would be relevant for applications that read SCIONPackets and not only the payload.
___
*[`pkg/snet/multihomed/outbound_addr.go` line 57 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZd3vSs18mFiBvifKhs:-OZnP8E4DkH-fyWPOUmz:b-a50evo) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/multihomed/outbound_addr.go#L57)):*
<details><summary><i>Previously, juagargi (Juan A. Garcia Pardo) wrote…</i></summary><blockquote>

Correct, your version is not only more elegant, is also locking less amount of time. I'm using it.
</blockquote></details>

Sorry, why is it necessary the size check outside? Now you don't have a read lock. In any case, the write lock that you acquire before the second check already locks for the right size before trying to write in it, doesn't it?



_Code snippet:_
```Go
	if len(remoteToEgress) < MaxAllowedCacheSize {
		// Beware of the RWMutex, it's read-locked already.
		muRemoteToEgress.Lock()
		// Check again to avoid race conditions. Could skipped it if exact size is not important.
		if len(remoteToEgress) < MaxAllowedCacheSize {
			remoteToEgress[remote] = egress
		}
		muRemoteToEgress.Unlock()
	}
```
___
*[`pkg/snet/snet.go` line 95 at r2](https://reviewable.io/reviews/juagargi/scion/53#-OZnNFgV03M70-3MGGmM:-OZnNFgV03M70-3MGGmN:b-kxu7v0) ([raw file](https://github.com/juagargi/scion/blob/0aced0a5ea7222c23d4c423d48b601ebcb6d62a3/pkg/snet/snet.go#L95)):*
> ```Go
> 
> // OpenRaw returns a PacketConn which listens on the specified address.
> // Nil or unspecified addresses are not supported.
> ```

Remove this comment


<!-- Sent from Reviewable.io -->

juagargi committed:
Remove no longer valid comment.

juagargi committed:
Remove not needed double locking of the cache in OutboundIP.

juagargi reviewed (commented):
Done. I've added your suggestion after the paragraph that starts with 
```
The detection of the changes should be done ...
```

> *[Reviewable](https://reviewable.io/reviews/juagargi/scion/53#-:-OZnx0SzBjj8GC2ZVdur:b7s091f)* status: 6 of 14 files reviewed, 3 unresolved discussions (waiting on @JordiSubira and @marcfrei)

___
*[`pkg/snet/reader.go` line 96 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZcJqjE2_808LP1zdxJ:-OZo2AqG7lUztEln8h7U:b8mjknd) ([raw file](https://github.com/juagargi/scion/blob/150bf5830b5efc85e77d81545aff2d58f76566e4/pkg/snet/reader.go#L96)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

If we are going to allow this, the check will even be better at `snet.OpenRaw()`.

This is not about the packet arriving, is about the information on the packet being meaningful, i.e., the src address on the SCION layer being decoupled from the underlay. So far, this wasn't the case for snet (we didn't allow for Unspecified address, not even before the dispatcher-less change). 

This may have implications, not all of them completely obvious perhaps at the moment, but I can try to come up with an "artifact" example: A SCION application that listens on 0.0.0.0 wants to display how many packets is receiving per interface. One could think that I can use pkt.Destination.Host (populated at snet.PacketConn.ReadFrom(pkt)) and infer from that address the interface for which I am receiving packets. This is not correct, because one could send a SCION packet to the underlay X.X.X.X with SCION destination address Y.Y.Y.Y. Of course in this model, "the attacker" is local (meaning within the AS). 

I am not saying we have to care about this "misuse case" which as I stated is a bit of an artifact, what I am saying is that we ever have this change on production, this has to be documented. My suggestion is even having it documented at snet.OpenRaw() because the example would be relevant for applications that read SCIONPackets and not only the payload.
</blockquote></details>

It makes sense to disallow cases like the one you mentioned.
I think the proper way would be t have a more solid mechanism of delivery (e.g. using the Linux kernel with a SCION socket, where I could bind my socket to my ISD-AS but no IP in particular for example, forcing reception of only SCION packets with destination my IA), instead of checking in `snet` the reasons of a correct delivery of the packet. But we don't have that delivery mechanism yet.
So after talking to you offline, we have doubts on whether we should always perform the check that the scion host address is one of our local interfaces or not.
Tagging @marcfrei as he has been working with NAT, and this behavior affects NAT on the receiver side.
___
*[`pkg/snet/snet.go` line 95 at r2](https://reviewable.io/reviews/juagargi/scion/53#-OZnNFgV03M70-3MGGmM:-OZo08Kg2MGApfoobtT4:b-896fix) ([raw file](https://github.com/juagargi/scion/blob/0aced0a5ea7222c23d4c423d48b601ebcb6d62a3/pkg/snet/snet.go#L95)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

Remove this comment
</blockquote></details>

Done.
___
*[`pkg/snet/multihomed/outbound_addr.go` line 57 at r1](https://reviewable.io/reviews/juagargi/scion/53#-OZd3vSs18mFiBvifKhs:-OZo0bQY3RSU3fzHKWNn:bcue7fo) ([raw file](https://github.com/juagargi/scion/blob/e9c911b34bf7502d175d6731fcc04e7ab2407baf/pkg/snet/multihomed/outbound_addr.go#L57)):*
<details><summary><i>Previously, JordiSubira wrote…</i></summary><blockquote>

Sorry, why is it necessary the size check outside? Now you don't have a read lock. In any case, the write lock that you acquire before the second check already locks for the right size before trying to write in it, doesn't it?
</blockquote></details>

Done.
I left it unintendedly because I'm a donkey. Now it locks to write, but only if the size is correct. If we consider that there is a huge number of concurrent threads, mostly likely to hit this code when the cache is already at `MaxAllowedCacheSize`, I could change the code to do a double lock:
1. Get the read lock.
2. Check the size, if the cache is already big enough, bail.
3. Because the size of the cache was not too big, release the read lock, acquire a write lock.
4. Check again to avoid races between other threads possibly adding to the cache between steps 2 and 3. 
5. Increment cache, release write lock.

Let me know if you think performance of the locking is critical under these conditions, and I'll change the code to the proposed steps.


<!-- Sent from Reviewable.io -->

juagargi reviewed (commented):
Question @JordiSubira you may know: how can I create a test that always runs in a docker container, with certain setup (multiple local interfaces, etc), even spawning more than one container and network?

> *[Reviewable](https://reviewable.io/reviews/juagargi/scion/53#-:-OZoF1xi31NqVKwtt4mz:b-1nks0m)* status: 6 of 14 files reviewed, 3 unresolved discussions (waiting on @JordiSubira and @marcfrei)



<!-- Sent from Reviewable.io -->

marcfrei commented:
As far as I can see, there's no negative interaction between this PR and the planned NAT address discovery (https://docs.scion.org/en/latest/dev/design/NAT-address-discovery.html). With NAT support, we would learn the SCION underlay source address via STUN, though. There will thus be no need to look up the outbound IP via multihomed.OutboundIP if STUN is used: https://github.com/jdslab/scion/blob/03154d48691cfc8cfd6d2cd3bc9af8a6b0cc0e35/demo/stun/test-client/main.go#L125

marcfrei commented on pkg/snet/multihomed/interfaces.go:
```diff
@@ -0,0 +1,176 @@
+)
+
+var (
+	remoteToEgress   map[netip.Addr]netip.Addr = make(map[netip.Addr]netip.Addr)
+	muRemoteToEgress sync.RWMutex              = sync.RWMutex{}
```
Wouldn't it be better to use sync.Map for this?

juagargi replied:
I just reviewed the new sync.Map that uses atomic, it is quite fast modulo the type assertion needed to get back the `netip.Addr`. All in full considered, it's >50% slower than a plain map and a r/w lock. I've added some benchmarks to the code to check this, please take a look.
These are my results:
```
goos: linux
goarch: amd64
pkg: github.com/scionproto/scion/pkg/snet/multihomed
cpu: Intel(R) Core(TM) i7-7700T CPU @ 2.90GHz
BenchmarkSyncMapWrites-8   	 2841009	       522.4 ns/op
BenchmarkSyncMapReads-8    	 4197412	       257.4 ns/op
BenchmarkMuMapWrites-8     	 4080162	       332.0 ns/op
BenchmarkMuMapReads-8      	 6577942	       215.5 ns/op
```

Whenever `sync.Map` supports generic types (which is planned according to https://github.com/golang/go/issues/71076) I would definitely use `sync.Map` instead, as its semantics are exactly the ones we need here and it's much less verbose than locking/unlocking every time.
<!-- Reviewable comment -O_zquua-69xiC3GvSAC:b8iyxku -->
<!-- Sent from Reviewable.io -->

marcfrei commented on pkg/snet/multihomed/outbound_addr.go:
```diff
@@ -0,0 +1,75 @@
+	// Check if the table contains an entry.
+	muRemoteToEgress.RLock()
+	egress, ok := remoteToEgress[remote]
+	muRemoteToEgress.RLocker().Unlock()
```
`muRemoteToEgress.RUnlock()`, right?

juagargi replied:
Done.
<!-- Reviewable comment -Oa-5Ko72nNi_nF6D_Fw:b-896fix -->
<!-- Sent from Reviewable.io -->

marcfrei commented on pkg/snet/multihomed/outbound_addr.go:
```diff
@@ -0,0 +1,75 @@
+
+	// Check if our cache is not too big already.
+	if len(remoteToEgress) < MaxAllowedCacheSize {
+		// Beware of the RWMutex, it's read-locked already.
```
Is this correct, didn't we unlock on line 37?

juagargi replied:
Done. It was a stale comment.
<!-- Reviewable comment -Oa-BT7yA-k7Q5AxOEyW:b-ss09x4 -->
<!-- Sent from Reviewable.io -->

marcfrei commented on pkg/snet/multihomed/outbound_addr.go:
```diff
@@ -0,0 +1,75 @@
+	egress, _ = netip.AddrFromSlice(eg)
+
+	// Check if our cache is not too big already.
+	if len(remoteToEgress) < MaxAllowedCacheSize {
```
Is this safe in the Go memory model (https://go.dev/ref/mem)? If yes, it would help me to have a comment explaining why. Do we need the optimization at all?

juagargi replied:
The code has now changed due to the very small performance optimization vs the introduced reasoning complexity. IIRC `go test -race` didn't complain, but let me answer the question anyway to check my sanity.
Is this safe? I think so, the rationale being that the first greater-than condition may be evaluated with a stale value, but it's okay because:
- It will be evaluated again with the synchronized value if it's less than the threshold.
- The `len(remoteToEgress)` value transitions only to equal or bigger, sequentially as time moves forward, regardless of the execution order of the threads (monotonicity). With this and the point above, we ensure that stale values are always checked when growing. If already equal to `MaxAllowedCacheSize` they are not synchronized, but it is not needed as they will _never_ decrease (monotonicity). This is similar to a double checked lock.
I hope this makes sense. Please check with me if there are mistakes here, as I keep on using these ideas nowadays.

Addendum: I'm leaving the text above to expose my train of thought. But the second point, monotonicity, is not true: there is the case where size decreases, namely when calling `invalidateAll`. This could make a thread skip the insertion of the IP address value into the map, which although not being extremely harmful, is not what I explained above.
<!-- Reviewable comment -Oa-7NWhAzMzA-I2hzi0:b-k3dh7e -->
<!-- Sent from Reviewable.io -->

marcfrei commented on pkg/snet/multihomed/interfaces.go:
```diff
@@ -0,0 +1,176 @@
+	// The local addresses are stored in an atomic pointer to allow tests to inspect the
+	// internal value of it without data races.
+	localAddresses = atomic.Pointer[[]netip.Addr]{}
+	ticker         = time.NewTicker(CheckInterfacesPeriod)
```
Could be local to `continuousCheckInterfaces`, couldn't it?

juagargi replied:
Done.
<!-- Reviewable comment -O_zoXXgCJ7wu7Bfq_sx:b-896fix -->
<!-- Sent from Reviewable.io -->

marcfrei commented on pkg/snet/multihomed/interfaces.go:
```diff
@@ -0,0 +1,176 @@
+	remoteToEgress   map[netip.Addr]netip.Addr = make(map[netip.Addr]netip.Addr)
+	muRemoteToEgress sync.RWMutex              = sync.RWMutex{}
+
+	// The local addresses are stored in an atomic pointer to allow tests to inspect the
+	// internal value of it without data races.
```
Is this added testability really worth the increased structural complexity?

juagargi replied:
Granted, it's horrible. I'm removing it, with the new changes this is not necessary.
<!-- Reviewable comment -Oa-EbWv9dtiO4JUZJrb:b-2kekxp -->
<!-- Sent from Reviewable.io -->

marcfrei commented on pkg/snet/multihomed/interfaces.go:
```diff
@@ -0,0 +1,176 @@
+	return addrs
+}
+
+func equalAddressList(a, b []netip.Addr) bool {
```
Use `slices.Equal` instead?

juagargi replied:
Done.
<!-- Reviewable comment -Oa-4qVV13n_FOMdVlJl:b-896fix -->
<!-- Sent from Reviewable.io -->

juagargi committed:
The ticker object is local to continuousCheckInterfaces.

juagargi committed:
Add benchmarks measuring sync.Map and RWMutex+map.

juagargi committed:
Use slices.Equal instead of custom function.

juagargi committed:
Use RUnlock() instead of RLocker().Unlock().

juagargi committed:
Fix comment.

juagargi committed:
Simplify storage of localAddresses to just a slice.

The atomic.Pointer seems not needed anymore, tests with -race work
alright without warnings.

juagargi committed:
Add comment clarifying interactions with NAT+STUN.

juagargi reviewed (commented):
Thanks Marc. If/when I open a PR against `scionproto` I will make sure to mention the interaction with NAT, so that whichever one gets merged second, can reuse the logic of finding out the egress IP addresses.
I have also added a comment in the `OutboundIP` public function referring to the interaction between NAT-STUN and `OutboundIP`

> *[Reviewable](https://reviewable.io/reviews/juagargi/scion/53#-:-Oa-OqcI25k8DgAWbVNz:b-jsrxt4)* status: 6 of 14 files reviewed, 10 unresolved discussions (waiting on @JordiSubira and @marcfrei)



<!-- Sent from Reviewable.io -->

<!-- Reviewable submission -Oa-Qw8QDdoKDs1ZmvVK -->

juagargi committed:
Fix build after rebasing on last master.

juagargi committed:
Move the unit test to an acceptance test.

juagargi committed:
Simplify test: server at 110, clients at 111 and 112.

Have one server sitting at 110, with two distinct IP addresses,
each suitable for one BR in 110.
Server listens in 0.0.0.0
Clients in 111 and 112 check the reply address corresponds to that
of the particular interface of the particular BR used to talk to the
server.

juagargi committed:
Remove deprecated unit test.

juagargi committed:
Linter findings: long lines and rand.Read in test without err check.

juagargi committed:
Linter: flake8 new line required.

juagargi committed:
More robust network attachment to server container.

juagargi committed:
Kill previous server processes.

juagargi committed:
Debug output added.

juagargi committed:
Fix: local IP depends on IPv4 vs IPv6.

juagargi committed:
More debug output, server side.

juagargi committed:
Server starts with nohup.

juagargi committed:
Fix log filepath and remote attempts loop.

juagargi committed:
WIP simplest test, just topology shape.

juagargi committed:
WIP Remove one interface.

juagargi committed:
WIP back to original test.

juagargi committed:
WIP simplify server flags.

juagargi committed:
WIP server listens forever, client runs three times.

Once for a bound server, then twice for an unbound server, sending
requests to the two different addresses.

juagargi committed:
WIP more logging.

juagargi committed:
WIP only diagnostics.

juagargi committed:
WIP diagnostics and packet capture

juagargi committed:
WIP diagnostics for the docker server IP address.

In CI, the packets arrive from the docker gateway, instead of the source
of the traffic (which is 111-BR1).

juagargi committed:
WIP diagnostics for docker network.

juagargi committed:
WIP force local_002 to be after scn_002 by explicitly appending it.

juagargi committed:
WIP diagnostics for routers' docker networks.

juagargi committed:
Back to testing.

juagargi committed:
Just test, remove diagnostics.

juagargi committed:
Cleanup.

juagargi committed:
Linter findings.
